### ISSUE TITLE
About the accuracy on base-to-new generalization setting

### ISSUE THREAD

**Issue by yxgnahz** [muzairkhattak/PromptSRC#3]
Hi,

Thanks very much for your interesting and inspiring work and well-organized codebase! I am trying to reproduce your results in the paper, and here are the results I obtained for the base-to-new generalization setting: 
 | Imagenet | Caltech101 | Food101 | Dtd | Ucf101 | Oxford_Flowers | Oxford_Pets | FGVC | Standford_Cars | Sun397 | Eurosat | AVG
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
Base | 77.60 | 98.07 | 90.67 | 83.53 | 86.93 | 98.00 | 95.17 | 42.87 | 78.43 | 82.53 | 92.83 | 84.24
Std | 0.14 | 0.12 | 0.09 | 0.92 | 0.05 | 0.14 | 0.21 | 0.88 | 0.25 | 0.12 | 0.97 |  
New | 70.30 | 93.87 | 91.57 | **59.63** | 78.93 | 77.10 | 97.33 | 36.60 | 75.30 | 78.80 | **72.93** | 75.67
Std | 0.08 | 0.05 | 0.09 | **3.52** | 2.07 | 0.14 | 0.49 | 0.42 | 0.59 | 0.41 | **3.92** |  
H-mean | 73.77 | 95.92 | 91.12 | 69.58 | 82.74 | 86.30 | 96.24 | 39.49 | 76.83 | 80.62 | 81.69 | 79.73

Basically, I find that the accuracies on base classes are consistent with the results in the paper and their variances are quite small. However, when it comes to the accuracies on the new classes, the accuracies are lower than that reported in the paper (on average 75.67 vs. 76.10). After carefully checking the results, I find that the gap mainly comes from two datasets, DTD and EuraSAT, where the variances are much larger than other classes. Therefore, I wonder whether this phenomenon is normal or not? Did you observe larger variances for the two classes during testing on new classes? 

Thanks in advance!

**Comment by muzairkhattak** [comment]
Hi @yxgnahz,

Thank you for showing interest in PromptSRC!

Regarding your query, we also observed a bit larger variances for these datasets, but the obtained results are same as mentioned in our main paper.

So I think, the reproduced performance should roughly match the main paper results. There can be slight variance in the results when reproducing results on different machines, but the shared results gap on average novel class performance is significant (75.67 vs 76.10). From your table, even for base classes, some datasets have different results, such as 78.43 % on Stanford cars vs reported 78.27 %.

In order to resolve the issue, can you kindly verify the following?

- You can make sure that you are using the same versions of PyTorch and other packages as mentioned in [our installation read-me files shown here](https://github.com/muzairkhattak/PromptSRC/blob/main/docs/INSTALL.md). We have noticed in our experiments that using different PyTorch and CUDA version leads to fluctuating results. So this might help in obtaining the exact results.

-  You may kindly verify if you are using the provided scripts for running your experiments and not changing any hyper-parameters in the config files. For training in base-to-novel setting, we use [this script](https://github.com/muzairkhattak/PromptSRC/blob/main/scripts/promptsrc/base2new_train.sh) which by default uses the `vit_b16_c2_ep20_batch4_4+4ctx` config file. 

- Optionally, can you also re-run your experiments using the commit version of `1af6e40b01e0c0c24fd8b6854be835209a0cde74` for the Dassl.pytorch library. I have noticed that we used slightly older version of the Dassl.pytorch. It should not be a problem, but you may just check this as well in-case.

I hope that would be helpful. Kindly let us know if this resolves your issue!

**Comment by yxgnahz** [comment]
Thanks for your prompt reply! I followed your installation guide to set up the environments (torch 1.9.0+cu111) and used exactly the same script you provided for the reproduction. All the experiments are conducted on an Nvidia 3090 GPU. I also tried with the provided Dassl version, and I found the results unchanged. Therefore, could you please show me the exact variances you observed for these datasets?

**Comment by muzairkhattak** [comment]
Hi @yxgnahz,

Thank you for verifying the installation.

Actually I do not have access to the machine where all the checkpoints and results log files are stored. I will try to get the access, and share the variance details here as soon as possible.

Lastly, can you clarify if you are training your models from the scratch to reproduce the numbers or are you using the provided pre-trained weights to reproduce the results?

Thank you!

**Comment by yxgnahz** [comment]
Hi @muzairkhattak ,

The results above are trained from scratch using the scripts you provided in the document. By the way, I just verified the official released checkpoint for base-to-new setting, and here are results (accuracy on new classes) I obtained by running the reproduce inference script:

DTD 63.08+/-3.15, Eurosat: 73.88+/-4.80

The averaged accuracies are very close to the results in the paper (less than 0.1%) and I think the minor gap may come from the difference between our hardware platform. Moreover, the variances from the released checkpoint are  just consistent with what I have got in the models trained from scratch, which indicates that these two datasets indeed have large variances on new classes.

**Comment by muzairkhattak** [comment]
Hi @yxgnahz,

Thank you for sharing this update. Yes, I think these two datasets have higher variance as discussed. 

If manageable for you, can you try to reproduce the results from scratch using another machine if that is available? This will be a double check. 

Thank you once again!

**Comment by muzairkhattak** [comment]
Hi @yxgnahz,

I have just verified reproducing the main paper results by training models from scratch using this repository. I am able to obtain roughly same results as mentioned in our paper. 

For example, the reproduced results for EuroSAT are **92.90%** for base classes and **73.83%** for novel classes. Similarly on DTD, the base class results are **83.53%** and novel class results are **62.97%**. 

To facilitate the process of reproducing the results on these datasets, I have provided the training and testing log files through [this link](https://drive.google.com/drive/folders/1knCXGwJF3H3arZe4YfcCCwCrr8T_-FfA?usp=sharing). You can verify and compare them with your log files.

I hope that would be helpful! 

Thank you and kind regards!

**Comment by yxgnahz** [comment]
Hi @muzairkhattak ,

Sorry I only have 3090 GPUs on my server. If I have chance, I will test this on another machine. I check the logs and I find all the settings are just the same. I believe the large variance leads to the performance gap in my reproduction and the value of the gap is normal considering the variance of the two datasets. 

Thanks for your discussion!

**Comment by muzairkhattak** [comment]
No problem! Thank you very much!
